package/gluon-core/luasrc/usr/lib/lua/gluon/util.lua

```diff
@@ -188,4 +190,78 @@ function M.log(message, verbose)
 	posix_syslog.syslog(posix_syslog.LOG_INFO, message)
 end
 
+-- Blindly close the file descriptors in a given table.
+local function close_fds(descriptors)
+	for _, fd in pairs(descriptors) do
+		posix_unistd.close(fd)
+	end
+end
+
+M.subprocess = {}
+
+M.subprocess.DEVNULL = -1
+M.subprocess.PIPE = 1
+
+-- Execute a program found using command PATH search, like the shell.
+-- Return the pid, as well as the I/O streams as pipes or nil on error.
+function M.subprocess.popen(path, argt, options)
+	local childfds = {}
+	local parentfds = {}
+	local stdiostreams = {stdin = 0, stdout = 1, stderr = 2}
+
+	for iostream in pairs(stdiostreams) do
+		if options[iostream] == M.subprocess.PIPE then
+			local piper, pipew = posix_unistd.pipe()
+			if iostream == "stdin" then
+				childfds[iostream] = piper
+				parentfds[iostream] = pipew
+			else
+				childfds[iostream] = pipew
+				parentfds[iostream] = piper
+			end
+		end
+	end
+
+	-- childfds: r0, w1, w2
+	-- parentfds: w0, r1, r2
+
+	local pid, errmsg, errnum = posix_unistd.fork()
+
+	if pid == nil then
+		close_fds(childfds)
+		close_fds(parentfds)
+		return nil, errmsg, errnum
+	elseif pid == 0 then
+		local null = -1
+		if M.contains(options, M.subprocess.DEVNULL) then
+			-- only open, if there's anything to discard
+			null = posix_fcntl.open('/dev/null', posix_fcntl.O_RDWR)
+		end
+
+		for iostream in pairs(stdiostreams) do
+			local option = options[iostream]
+			if option == M.subprocess.DEVNULL then
+				posix_unistd.dup2(null, stdiostreams[iostream])
+			elseif option == M.subprocess.PIPE then
+				-- only close these, if they exist
+				posix_unistd.close(parentfds[iostream])
+				posix_unistd.dup2(childfds[iostream], stdiostreams[iostream])
+				-- close all the dups
+				posix_unistd.close(childfds[iostream])
+			end
+		end
```

Comment:
I think the following is be a bit easier to understand:
```suggestion
		for iostream, fd in pairs(stdiostreams) do
			local option = options[iostream]
			if option == M.subprocess.DEVNULL then
				posix_unistd.dup2(null, fd)
			elseif option == M.subprocess.PIPE then
				posix_unistd.dup2(childfds[iostream], fd)
			end
		end
		close_fds(childfds)
		close_fds(parentfds)
```

Reply:
I think the closing should stay in the PIPE-branch, as it's not necessary in the others, right?
Everything else looks good.

Reply:
Well, as long as it does no harm, I'm fine with both.

Reply:
resolved in 1acc783 and 98efb82